---
title: "CD45+Thy1+ ILC hdWGCNA analysis"
output:
   html_document: 
     theme: united
     df_print: kable
     toc: yes
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include = FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```


# Universal packages

```{r Load packages}
library(Seurat)
library(tidyverse)
library(patchwork)
library(cowplot)
```


# Data input

```{r Load SCT processed data}
# Load CD45+Thy1+ data

# Thy1+
Thy1 <- readRDS("CD45_Thy1_DP_labeled_paper_ver.rds")


# Set RNA assay as default
DefaultAssay(Thy1) <- "RNA"
head(x = Thy1[[]])
```


# CD45+ Thy1+ cluster Visualization

```{r Set Reveal CD45+Thy1+ clusters}
# Show the numbers for CD45+Thy1+
table(Thy1@meta.data$cell_type, split.by = Thy1@meta.data$treatment)

# Plot the CD45+ Thy1+ clusters
DimPlot(Thy1, group.by = "cell_type", label = T, repel =  T)
DimPlot(Thy1, group.by = "cell_type", split.by = "treatment", label = F) + ggtitle(NULL)
```


# Overall Thy1 subsets

```{r Define cell subsets}
# Label the overall annotation
overall.ids.CNA <- c("ILC2", "ILC2", "ILC2", "ILC2", "ILC2", "ILC2", "T", "NKL", "ILC1L", "T", "HSC", "NKL", "T")
names(overall.ids.CNA) <- levels(Thy1)
Thy1 <- RenameIdents(Thy1, overall.ids.CNA)
Thy1 <- AddMetaData(Thy1, metadata = Thy1@active.ident, col.name = 'cell_type_overall')

# Show the numbers across the conditions
table(Thy1@active.ident, split.by = Thy1@meta.data$treatment)

# Plot the CD45+Thy1+ clusters
DimPlot(Thy1, group.by = "cell_type", label = T, repel =  T)
DimPlot(Thy1, group.by = "cell_type", split.by = "treatment", label = F) + ggtitle(NULL)
```


# Packages for hdWGCNA and GSEA

```{r Prepare the packages for hdWGCNA}
# Load co-expression network analysis packages
library(WGCNA)
library(hdWGCNA)

# Load network analysis & visualization package
library(igraph)
```

```{r Prepare the packages for GSE and pathway analysis}
# Load packages for enrichment analysis (enrichR)
library(enrichR)

# Check available databases to perform enrichment (then choose one)
enrichR::listEnrichrDbs()
```

```{r Prepare the packages for GeneOverlap}
# Marker gene overlap analysis
library(GeneOverlap)
```


# Preparation of co-expression network analysis (CNA) on overall ILC types

```{r Prepare to run overall ILC coexpression network}
# Set up Seurat object for WGCNA for overall ILC
Thy1 <- SetupForWGCNA(Thy1,
  gene_select = "fraction",
  fraction = 0.05, 
  # fraction of cells that a gene needs to be expressed in order to be included
  wgcna_name = "Overall ILC hdWGCNA" 
  # the name of the hdWGCNA experiment
)

# Construct metacells  in each group
Thy1 <- MetacellsByGroups(
  Thy1,
  group.by = c("cell_type_overall", "treatment"),
  # specify the columns in seurat_obj@meta.data to group by
  reduction = 'umap',
  # select the dimensionality reduction to perform KNN on
  assay = 'RNA',
  slot = 'counts',
  k = 25,
  # nearest-neighbors parameter
  max_shared = 10,
  # maximum number of shared cells between two metacells
  ident.group = 'cell_type_overall'
  # set the Idents of the metacell seurat object
)

# Normalize metacell expression matrix
Thy1 <- NormalizeMetacells(Thy1)
```


# ILC Alt.BPC hdWGCNA

```{r Run ILC_BPC coexpression network}
# Set up the expression matrix on ILC
ILC.BPC_CNA <- SetDatExpr(
  Thy1,
  group.by='cell_type_overall',
  # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups
  group_name = c("NKL", "ILC1L", "ILC2"),
  # the name of the group of interest in the group.by column
  multi.group.by = 'treatment',
  #A string containing the name of a column in the Seurat object with groups for consensus WGCNA (dataset, sample, condition, etc)
  multi_group_name = c("Alt.BPC"),
  #A string containing the name of a group present in the multi.group.by column
  assay = 'RNA',
  # using RNA assay
  slot = 'data'
  # using normalized data
)

# Test different soft powers
ILC.BPC_CNA <- TestSoftPowers(
  ILC.BPC_CNA,
  networkType = 'signed'
  # "unsigned" or "signed hybrid" can also be used
)

# Plot the results
plot_list_ILC <- PlotSoftPowers(ILC.BPC_CNA)

# Reveal the data
wrap_plots(plot_list_ILC, ncol = 2)

power_table_ILC <- GetPowerTable(ILC.BPC_CNA)
head(power_table_ILC)

# Construct co-expression network
ILC.BPC_CNA <- ConstructNetwork(
  ILC.BPC_CNA,
  setDatExpr = FALSE,
  overwrite_tom = TRUE,
  tom_name = 'ILC_BPC-log'
  # name of the topoligical overlap matrix written to disk
)

PlotDendrogram(ILC.BPC_CNA, main='ILC Alt.BPC hdWGCNA Dendrogram')

# Need to run ScaleData first or else harmony throws an error
ILC.BPC_CNA <- ScaleData(ILC.BPC_CNA, features=VariableFeatures(ILC.BPC_CNA))

# Compute all MEs in the full single-cell dataset
ILC.BPC_CNA <- ModuleEigengenes(ILC.BPC_CNA)

# Compute eigengene-based connectivity (kME)
ILC.BPC_CNA <- ModuleConnectivity(ILC.BPC_CNA)

# Rename the modules
ILC.BPC_CNA <- ResetModuleNames(
  ILC.BPC_CNA,
  new_name = "M"
)

# Get the module assignment table:
ILC.BPC_modules <- GetModules(ILC.BPC_CNA)

# show the first 10 columns:
head(ILC.BPC_modules[,1:10])
```

```{r Plot ILC_BPC coexpression network, fig.height = 4, fig.width = 10}
# Plot genes raILCed by kME for each module
PlotKMEs(ILC.BPC_CNA, ncol = 5)
```

```{r Change the colors of selected modules}
# Set new colors
new_module_colors <- list('M5' = 'darksalmon')
ILC.BPC_CNA <- ResetModuleColors(ILC.BPC_CNA, new_module_colors )
```

```{r Plot ILC_BPC hub gene network, fig.height = 10, fig.width = 10}
# Plot hubgene network
HubGeneNetworkPlot(
  ILC.BPC_CNA,
  n_hubs = 6, n_other = 0,
  edge_prop = 1,
  edge.alpha = 0.3,
  vertex.label.cex = 1,
  hub.vertex.size = 5,
  other.vertex.size = 0,
  mods = 'all'
)
```

```{r ILC_BPC Enrichment analysis}
# Perform enrichment tests
ILC.BPC_CNA <- RunEnrichr(
  ILC.BPC_CNA,
  dbs = "WikiPathways_2019_Mouse",
  # character vector of enrichr databases to test
  max_genes = 100 
  # number of genes per module to test. use max_genes = Inf to choose all genes!
)

# Retrieve the output table
GetEnrichrTable(ILC.BPC_CNA)
```

```{r ILC_BPC EnrichrDotPlot for paper, fig.width = 7, fig.height = 5}
# Enrichr dotplot
ILC.EnrichrDotPlot <- EnrichrDotPlot(
  ILC.BPC_CNA,
  mods = "all",
  # use all modules (this is the default behavior)
  database = "WikiPathways_2019_Mouse",
  # this has to be one of the lists we used above!!!
  n_terms=2 # number of terms for each module
) + theme(axis.text.x = element_text(size = 12, angle = 0, vjust = 0.5, hjust = 0.5)) +
  ggtitle("Enrichment Analysis")

ILC.EnrichrDotPlot
```

```{r Rename the modules for paper, fig.height = 7, fig.width = 14}
# Rename the modules
rename_list <- list(
  "M1" = "M1: TYROBP causal network",
  "M2" = "M2: DNA replication",
  "M3" = "M3: MAPK signaling pathway",
  "M4" = "M4: Type II interferon signaling (IFNG)",
  "M5" = "M5: Cytoplasmic ribosomal proteins",
  "M6" = "M6: p53 signaling",
  "M7" = "M7: Neuclear receptors (Rora-related)",
  "M8" = "M8: Electron transport chain",
  "M9" = "M9: Cytoplasmic ribosomal proteins"
)

ILC.BPC_CNA_renamed <- ResetModuleNames(
  ILC.BPC_CNA,
  new_name = rename_list
)

# Plot genes raILCed by kME for each module
ILC.PlotKMEs <- PlotKMEs(ILC.BPC_CNA_renamed,
                         ncol = 3,
                         text_size = 3,
                         plot_widths = c(1, 3)
                         )

ILC.PlotKMEs

# Get the module assignment table:
modules <- GetModules(ILC.BPC_CNA_renamed)

# Show the first 4 columns:
head(modules[,1:4])
```

```{r ILC_BPC coexpression dotplot, fig.width = 5.5, fig.height = 6}
# Compute gene scoring for the top 25 hub genes by kME for each module
# With Seurat method
ILC.BPC_CNA <- ModuleExprScore(
  ILC.BPC_CNA,
  n_genes = 25,
  method='Seurat'
)

# Get hMEs from seurat object
ILC.BPC_MEs <- GetMEs(ILC.BPC_CNA, harmonized=TRUE)
ILC.BPC_mods <- colnames(ILC.BPC_MEs); ILC.BPC_mods <- ILC.BPC_mods[ILC.BPC_mods != 'grey']

# Reorder the vector ILC.BPC_mods into numeric order
ILC.BPC_mods <- sort(ILC.BPC_mods)


# Add hMEs to Seurat meta-data:
ILC.BPC_CNA@meta.data <- cbind(ILC.BPC_CNA@meta.data, ILC.BPC_MEs)

# Plot with Seurat's DotPlot function
DotPlot(ILC.BPC_CNA, features = ILC.BPC_mods, group.by = 'cell_type', split.by = 'treatment', cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(size = 10))
```

```{r Set bqck primary ident}
# Set the primary ident to "cell_type"
ILC.BPC_CNA <- SetIdent(ILC.BPC_CNA, value = "cell_type")
```

```{r ILC_BPC marker gene OverlapDotPlot, fig.width = 5, fig.height = 3}
# Compute marker genes in ILC with Seurat
ILC.BPC_CNA_ILC <- ILC.BPC_CNA %>%
  subset(idents = c("Tcm", "Teff", "HSC", "Tem"), invert = TRUE)

table(ILC.BPC_CNA_ILC@meta.data$cell_type, split.by = ILC.BPC_CNA_ILC@meta.data$treatment)

ILC.BPC_ILC_markers <- Seurat::FindAllMarkers(
  ILC.BPC_CNA_ILC,
  only.pos = TRUE,
  logfc.threshold=1
)

# Compute marker gene overlaps in ILC
ILC.BPC_ILC_overlap <- OverlapModulesDEGs(
  ILC.BPC_CNA_ILC,
  deg_df = ILC.BPC_ILC_markers,
  fc_cutoff = 1 # log fold change cutoff for overlap analysis
)

# Plot odds ratio of the overlap as a dot plot for ILC
ILC_OverlapDotPlot <- OverlapDotPlot(
  ILC.BPC_ILC_overlap,
  plot_var = 'odds_ratio') +
  ggtitle('Overlap of ILC_BPC modules to cell types') + 
  theme(axis.text.x = element_text(size = 10, angle = 0, vjust = 0.5, hjust = 0.5))

ILC_OverlapDotPlot
```

```{r ILC_BPC marker gene in NK/ILC1 OverlapDotPlot, fig.width = 5, fig.height = 2}
# Compute marker genes in NK/ILC1 with Seurat
ILC.BPC_CNA_ILC_NK <- ILC.BPC_CNA %>%
  subset(idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), invert = FALSE) %>%
  SetIdent(value = "treatment")

table(ILC.BPC_CNA_ILC_NK@meta.data$cell_type_overall, split.by = ILC.BPC_CNA_ILC_NK@meta.data$treatment)

ILC.BPC_ILC1_NK_markers <- Seurat::FindAllMarkers(
  ILC.BPC_CNA_ILC_NK,
  only.pos = TRUE,
  logfc.threshold=0.5
)

# Compute marker gene overlaps in NK/ILC1
ILC.BPC_ILC1_NK_overlap <- OverlapModulesDEGs(
  ILC.BPC_CNA_ILC_NK,
  deg_df = ILC.BPC_ILC1_NK_markers,
  fc_cutoff = 0.5 # log fold change cutoff for overlap analysis
)

# Plot odds ratio of the overlap as a dot plot for NK/ILC1
ILC1_NK_OverlapDotPlot <- OverlapDotPlot(
  ILC.BPC_ILC1_NK_overlap,
  plot_var = 'odds_ratio') +
  ggtitle('Overlap of modules in NK/ILC1') + 
  theme(axis.text.x = element_text(size = 10, angle = 0, vjust = 0.5, hjust = 0.5))

ILC1_NK_OverlapDotPlot
```

```{r ILC_BPC marker gene in NK/ILC1 OverlapBarPlot, fig.width = 3, fig.height = 3}
# Overlap barplot for Alt, produces a plot for each cell type
plot_list_ILC1_NK <- OverlapBarPlot(ILC.BPC_ILC1_NK_overlap)

# Stitch plots with patchwork
ILC1_NK_OverlapBarPlot <- wrap_plots(plot_list_ILC1_NK)

ILC1_NK_OverlapBarPlot.title <- plot_grid(NULL, ILC1_NK_OverlapBarPlot, labels = c('NK/ILC1'), label_size = 12, ncol = 1, rel_heights = c(0.07, 1))

ILC1_NK_OverlapBarPlot.title
```

```{r ILC_BPC marker gene in ILC2 OverlapDotPlot, fig.width = 5, fig.height = 2}
# Compute marker genes in ILC2 with Seurat
ILC.BPC_CNA_ILC2 <- ILC.BPC_CNA %>%
  subset(idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), invert = FALSE) %>%
  SetIdent(value = "treatment")

table(ILC.BPC_CNA_ILC2@meta.data$cell_type_overall, split.by = ILC.BPC_CNA_ILC2@meta.data$treatment)

ILC.BPC_ILC2_markers <- Seurat::FindAllMarkers(
  ILC.BPC_CNA_ILC2,
  only.pos = TRUE,
  logfc.threshold=0.5
)

# Compute marker gene overlaps in ILC2
ILC.BPC_ILC2_overlap <- OverlapModulesDEGs(
  ILC.BPC_CNA_ILC2,
  deg_df = ILC.BPC_ILC2_markers,
  fc_cutoff = 0.5 # log fold change cutoff for overlap analysis
)

# Plot odds ratio of the overlap as a dot plot for ILC2
ILC2_OverlapBarPlot <- OverlapDotPlot(
  ILC.BPC_ILC2_overlap,
  plot_var = 'odds_ratio') +
  ggtitle('Overlap of modules in ILC2') + 
  theme(axis.text.x = element_text(size = 10, angle = 0, vjust = 0.5, hjust = 0.5))

ILC2_OverlapBarPlot
```

```{r ILC_BPC marker gene in ILC2 OverlapBarPlot, fig.width = 3, fig.height = 3}
# Overlap barplot for Alt, produces a plot for each cell type
plot_list_ILC2 <- OverlapBarPlot(ILC.BPC_ILC2_overlap)

# Stitch plots with patchwork
ILC2_OverlapBarPlot <- wrap_plots(plot_list_ILC2)

ILC2_OverlapBarPlot.title <- plot_grid(NULL, ILC2_OverlapBarPlot, labels = c('    ILC2'), label_size = 12, ncol = 1, rel_heights = c(0.07, 1))

ILC2_OverlapBarPlot.title
```


# ILC_BPC module gene sets

```{r ILC_BPC module gene sets}
# ILC_BPC module genes fromhdWGCNA
ILC_BPC_M1 <- c("Tyrobp", "Plek", "Ugcg", "Gzma", "Fcer1g", "Klrd1", "Car2", "Klrk1", "Nkg7", "Ms4a4b")

ILC_BPC_M2 <- c("Birc5", "Ptma", "Rrm1", "4921515E04Rik", "Tubb5", "Rrm2", "Hmgb2", "Stmn1", "Tuba1b", "Il31ra")

ILC_BPC_M3 <- c("Lmo4", "Sik1", "Ywhaz", "Zc3hav1", "Psap", "Nr4a3", "Ddx3x", "Dusp1", "Fosb", "Nr4a1")

ILC_BPC_M4 <- c("Gbp6", "Gbp2", "Tgtp2", "Samhd1", "Zbp1", "Slfn5", "Ifit1", "Stat1", "Rsad2", "Isg15")

ILC_BPC_M5 <- c("Rp135", "Rps18", "Rps28", "Rps27", "Gm10076", "Rps19", "Rps5", "Rpl23a", "Rpl37a", "Rpl41")

ILC_BPC_M6 <- c("Sdf4", "Axl", "Timp2", "Pmepa1", "Tff1", "Tnfsf11", "Eef2", "Ndfip1", "Edn3", "AA467197")

ILC_BPC_M7 <- c("Pcsk1", "Cd82", "Cish", "Areg", "Il1rl1", "Gadd45b", "Csf2", "Bhlhe40", "Furin", "Gata3")

ILC_BPC_M8 <- c("Cox6a1", "Uqcrq", "Calm1", "Gm53028", "Uqcr10", "Atp5e", "Oaz1", "Cfap77", "Gapdh", "Lgals1")

ILC_BPC_M9 <- c("Rplp1", "Rps21", "Rps10", "Rps20", "Rpsa", "Rps27a", "Rpl38", "Rps4x", "Rps15a", "Rps12")


Module_M1_M3_M7 <- list("M1" = ILC_BPC_M1, "M3" = ILC_BPC_M3, "M7" = ILC_BPC_M7)

Module_M2_M6 <- list("M2" = ILC_BPC_M2, "M6" = ILC_BPC_M6)

Module_M5_M9 <- list("M5" = ILC_BPC_M5, "M9" = ILC_BPC_M9)


Module_M4 <- list(" " = ILC_BPC_M4)

Module_M8 <- list(" " = ILC_BPC_M8)
```


# NK/ILC1 module gene expression

```{r ILC_BPC module expression in NK_ILC1, fig.width = 5, fig.height = 3}
# Plot with Seurat's DotPlot function
NK_ILC1_module <- DotPlot(ILC.BPC_CNA, features = ILC.BPC_mods, idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = 'cell_type', split.by = 'treatment', cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(size = 10, color = c("dimgray", "black"))) + ggtitle("Module expression")

NK_ILC1_module
```

```{r NK_ILC1 module expression, fig.width = 6, fig.height = 4}
# Dotplot
DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M1), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M1: TYROBP causal network")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M2), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M2: DNA replication")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M3), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M3: MAPK signaling pathway")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M4), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M4: Type II interferon signaling (IFNG)")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M5), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M5: Cytoplasmic ribosomal proteins")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M6), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M6: p53 signaling")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M7), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M7: Neuclear receptors (Rora-related)")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M8), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M8: Electron transport chain")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M9), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M9: Cytoplasmic ribosomal proteins")
```

```{r ILC.BPC_CNA module expression for paper, fig.width = 11, fig.height = 4}
NK_ILC1_M1_M3_M7 <- DotPlot(ILC.BPC_CNA, features = Module_M1_M3_M7, idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M1: TYROBP causal network/M3: MAPK signaling pathway/M7: Neuclear receptors (Rora-related)")

NK_ILC1_M1_M3_M7


NK_ILC1_M2_M6 <- DotPlot(ILC.BPC_CNA, features = Module_M2_M6, idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M2: DNA replication/M6: p53 signaling")

NK_ILC1_M2_M6


NK_ILC1_M8 <- DotPlot(ILC.BPC_CNA, features = Module_M8, idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M8: Electron transport chain")

NK_ILC1_M8


NK_ILC1_M5_M9 <- DotPlot(ILC.BPC_CNA, features = Module_M5_M9, idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M5/M9: Cytoplasmic ribosomal proteins")

NK_ILC1_M5_M9


NK_ILC1_M4 <- DotPlot(ILC.BPC_CNA, features = Module_M4, idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M4: Type II interferon signaling (IFNG)")

NK_ILC1_M4
```


# ILC2 module expression

```{r ILC_BPC module expression in ILC2, fig.width = 5, fig.height = 5}
# Plot with Seurat's DotPlot function
ILC2_module <- DotPlot(ILC.BPC_CNA, features = ILC.BPC_mods, idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = 'cell_type', split.by = 'treatment', cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(size = 10, color = c("dimgray", "black"))) + ggtitle("Module expression")

ILC2_module
```

```{r ILC2 module expression, fig.width = 7, fig.height = 5}
# Dotplot
DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M1), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M1: TYROBP causal network")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M2), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M2: DNA replication")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M3), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M3: MAPK signaling pathway")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M4), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M4: Type II interferon signaling (IFNG)")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M5), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M5: Cytoplasmic ribosomal proteins")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M6), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M6: p53 signaling")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M7), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M7: Neuclear receptors (Rora-related)")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M8), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M8: Electron transport chain")

DotPlot(ILC.BPC_CNA, features = rev(ILC_BPC_M9), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) + ggtitle("M9: Cytoplasmic ribosomal proteins")
```

```{r ILC2 module expression for paper, fig.width = 11, fig.height = 5}
ILC2_M1_M3_M7 <- DotPlot(ILC.BPC_CNA, features = Module_M1_M3_M7, idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M1: TYROBP causal network/M3: MAPK signaling pathway/M7: Neuclear receptors (Rora-related)")

ILC2_M1_M3_M7


ILC2_M2_M6 <- DotPlot(ILC.BPC_CNA, features = Module_M2_M6, idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M2: DNA replication/M6: p53 signaling")

ILC2_M2_M6


ILC2_M8 <- DotPlot(ILC.BPC_CNA, features = Module_M8, idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M8: Electron transport chain")

ILC2_M8


ILC2_M5_M9 <- DotPlot(ILC.BPC_CNA, features = Module_M5_M9, idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M5/M9: Cytoplasmic ribosomal proteins")

ILC2_M5_M9


ILC2_M4 <- DotPlot(ILC.BPC_CNA, features = Module_M4, idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("M4: Type II interferon signaling (IFNG)")

ILC2_M4
```


# Oxidative stress gene sets

```{r Functional gene sets}
# Stress genes
Stress <- c("Grpe1", "Grpe2", "Dnajb1", "Hspd1", "Hsp1a1", "Hspa5", "Hspa8", "Hsp90aa1", "Hsp90ab1", "Hsp90b1")

# Antioxidant genes
AntiOx <- c("Gpx1", "Gpx4", "ndufa13", "Prdx1", "Prdx2", "Prdx3", "Prdx4", "Prdx5", "Prdx6", "Sod1")
```


# Oxidative stress gene expression in NK and ILC1

```{r NK_ILC1 Stressidant genes, fig.width = 6, fig.height = 3}
NK_ILC1_Stress <- DotPlot(ILC.BPC_CNA, features = c(Stress), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("Stress-related genes")

NK_ILC1_Stress
```

```{r NK_ILC1 Antioxidant genes, fig.width = 5, fig.height = 3}
NK_ILC1_AntiOx <- DotPlot(ILC.BPC_CNA, features = c(AntiOx), idents = c("NCR1 NK/ILC1", "IFNg NK/ILC1", "ST2 NK/ILC1"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("Antioxidant genes")

NK_ILC1_AntiOx
```


# Oxidative stress gene expression in ILC2

```{r ILC2 Stressidant genes, fig.width = 7, fig.height = 5}
ILC2_Stress <- DotPlot(ILC.BPC_CNA, features = c(Stress), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("Stress-related genes")

ILC2_Stress
```

```{r ILC2 Antioxidant genes, fig.width = 6, fig.height = 5}
ILC2_AntiOx <- DotPlot(ILC.BPC_CNA, features = c(AntiOx), idents = c("Areg ILC2", "IL-13 ILC2", "IL-10 ILC2", "Quiescent ILC2", "KLRG1 ILC2", "LTB ILC2"), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("Antioxidant genes")

ILC2_AntiOx
```


# Data oupput

```{r ILC_BPC figure for paper, fig.width = 12, fig.height = 15}
# Generate the plot in paper figure form
figure4_1 <- plot_grid(NULL, ILC.PlotKMEs, labels = c('(A)'), label_size = 18, nrow = 1, ncol = 2, rel_widths = c(0.2, 2))
figure4_2 <- plot_grid(ILC.EnrichrDotPlot, labels = c('(B)'), label_size = 18, ncol = 1)
figure4_3 <- plot_grid(NULL, NULL, NULL, NULL, labels = c('(C)', 'NK/ILC1', '(D)', 'ILC2'), label_size = 18, nrow = 1, rel_widths = c(0.72, 1.27, 0.87, 1.13))
figure4_4 <- plot_grid(ILC1_NK_OverlapBarPlot, ILC2_OverlapBarPlot)

figure4 <- plot_grid(figure4_1, figure4_2, figure4_3, figure4_4,
                     nrow = 4,
                     ncol = 1, 
                     rel_heights = c(1.4, 1, 0.1, 0.6)
                     )

figure4
```

```{r Module expression figure for paper, fig.width = 16, fig.height = 9}
# Generate the plot in paper figure form
figure5_1 <- plot_grid(NULL, NK_ILC1_module, NULL, NK_ILC1_Stress, NULL, NK_ILC1_AntiOx, labels = c('(A)', ' ', '(C)', ' ', '(E)'), label_size = 18, nrow = 1, rel_widths = c(0.1, 1, 0.1, 1, 0.1, 1))
figure5_2 <- plot_grid(ILC2_module, ILC2_Stress, ILC2_AntiOx, labels = c('(B)', '(D)', '(F)'), label_size = 18, nrow = 1, rel_widths = c(1, 1, 1))

figure5 <- plot_grid(figure5_1, figure5_2,
  nrow = 2,
  ncol = 1,
  rel_heights = c(1, 1.2)
  )

figure5
```

```{r NK_ILC1 figure for paper, fig.width = 14, fig.height = 12}
# Generate the plot in paper figure form
figureS3_1 <- plot_grid(NULL, NK_ILC1_M1_M3_M7, labels = c('(A)'), label_size = 18, nrow = 1, rel_widths = c(0.01, 2.3))
figureS3_2 <- plot_grid(NK_ILC1_M2_M6, NK_ILC1_M8, labels = c('(B)', '(C)'), label_size = 18, nrow = 1, rel_widths = c(1.4, 1))
figureS3_3 <- plot_grid(NK_ILC1_M4, NULL, NK_ILC1_M5_M9, labels = c('(D)', ' ', '(E)'), label_size = 18, nrow = 1, rel_widths = c(1, 0.1, 1.5))

figureS3 <- plot_grid(figureS3_1, figureS3_2, figureS3_3,
  nrow = 3,
  ncol = 1,
  rel_heights = c(1, 1.1, 1)
  )

figureS3
```

```{r ILC2 figure for paper, fig.width = 14, fig.height = 14}
# Generate the plot in paper figure form
figureS4_1 <- plot_grid(NULL, ILC2_M1_M3_M7, labels = c('(A)'), label_size = 18, nrow = 1, rel_widths = c(0.01, 2.3))
figureS4_2 <- plot_grid(ILC2_M2_M6, ILC2_M8, labels = c('(B)', '(C)'), label_size = 18, nrow = 1, rel_widths = c(1.3, 1))
figureS4_3 <- plot_grid(ILC2_M4, NULL, ILC2_M5_M9, labels = c('(D)', ' ', '(E)'), label_size = 18, nrow = 1, rel_widths = c(1, 0.1, 1.5))

figureS4 <- plot_grid(figureS4_1, figureS4_2, figureS4_3,
  nrow = 3,
  ncol = 1,
  rel_heights = c(1, 1.1, 1.1)
  )

figureS4
```


# Data save

```{r Output the figures}
# Create a folder for the figures
dir.create("./Figures", showWarnings = FALSE)

# Save the figure
ggsave2(plot = figure4, path = "./Figures", "Figure4.tiff", device = "tiff", dpi = 300, bg = "white", width = 12, height = 15, units = "in")
ggsave2(plot = figure5, path = "./Figures", "Figure5.tiff", device = "tiff", dpi = 300, bg = "white", width = 16, height = 9, units = "in")
ggsave2(plot = figureS3, path = "./Figures", "figureS3.tiff", device = "tiff", dpi = 300, bg = "white", width = 14, height = 12, units = "in")
ggsave2(plot = figureS4, path = "./Figures", "figureS4.tiff", device = "tiff", dpi = 300, bg = "white", width = 14, height = 14, units = "in")
```

```{r Output the CSV files}
# Create a folder for CSV files
dir.create("./CSV", showWarnings = FALSE)

# Export to csv
write.csv(modules, file.path("./CSV", "ILC.BPC_hdWGCNA_modules.csv"))
```


<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```

