---
title: "CD45+Thy1+ ILC TF activity analysis and HSP expression"
output:
   html_document: 
     theme: united
     df_print: kable
     toc: yes
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include = FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```


# Universal packages

```{r Load packages}
library(Seurat)
library(tidyverse)
library(patchwork)
library(cowplot)
```


# Data input (CD45+Thy1- and CD45-)

```{r Load SCT processed Ptprc and Structural data}
# Load Sensitivity 2 subsetted data

# CD45+ Thy1-
Ptprc <- readRDS("CD45_Thy1_neg_labeled_paper_ver.rds")

# CD45-
Structural <- readRDS("CD45_neg_labeled_paper_ver.rds")

# ".." is a special directory name meaning “the directory above the current one” or the parent directory

# Set RNA assay as default
DefaultAssay(Ptprc) <- "RNA"
head(x = Ptprc[[]])

DefaultAssay(Structural) <- "RNA"
head(x = Structural[[]])
```


# CD45+ Thy1- cluster Visualization

```{r Reveal CD45+ Thy1- clusters}
# Show the numbers for CD45+ Thy1-
table(Ptprc@meta.data$cell_type, split.by = Ptprc@meta.data$treatment)

# Plot the CD45+ Thy1- clusters
DimPlot(Ptprc, group.by = "cell_type", label = T, repel =  T)
DimPlot(Ptprc, group.by = "treatment", label = F) + ggtitle(NULL)
```


# CD45- cluster Visualization

```{r Reveal CD45- clusters}
# Show the numbers for CD45-
table(Structural@meta.data$cell_type, split.by = Structural@meta.data$treatment)

# Plot the CD45- clusters
DimPlot(Structural, group.by = "cell_type", label = T, repel =  T)
DimPlot(Structural, group.by = "treatment", label = F) + ggtitle(NULL)
```


# Packages for volcano plot

```{r Load the package for volcano plot}
# Load the package for volcano plot
library(EnhancedVolcano)
```


# CD45+ Thy1- differential expression

```{r Ptprc differential expression}
Ptprc <- SetIdent(Ptprc, value = "treatment")

# Find the differential expressed genes in Ptprc
Ptprc_differential <- FindMarkers(Ptprc, ident.1 = "Alt.BPC", ident.2 = "Alt", min.cells.group = 10, test.use = "wilcox_limma", only.pos = FALSE, assay = "RNA", slot = "scale.data")

Ptprc_differential <- Ptprc_differential[order(abs(Ptprc_differential$avg_log2FC), decreasing = T), ]

head(Ptprc_differential)

# Find the top 5 (positive and negative) differential expression in ILC2
Ptprc_differential %>%
  filter(abs(avg_log2FC) > 0.5) %>%
  filter(avg_log2FC > 0) %>%
  top_n(-5, p_val) %>%
  arrange(p_val) -> Ptprc_differential_top5pos

head(Ptprc_differential_top5pos)

Ptprc_differential %>%
  filter(abs(avg_log2FC) > 0.5) %>%
  filter(avg_log2FC < 0) %>%
  top_n(-5, p_val) %>%
  arrange(p_val) -> Ptprc_differential_top5neg

head(Ptprc_differential_top5neg)
```

```{r Ptprc DE Volcano plot, fig.width = 6, fig.height = 6}
# Plot the top 10 differential genes with volcano plot
Ptprc_volcano <- EnhancedVolcano(Ptprc_differential,
    lab = rownames(Ptprc_differential),
    selectLab = c(rownames(Ptprc_differential_top5pos), rownames(Ptprc_differential_top5neg)),
    x = 'avg_log2FC',
    y = 'p_val',
    title = 'CD45+Thy1- Alt.BPC vs. Alt',
    pCutoff = 10e-10,
    FCcutoff = 0.5,
    pointSize = 3.0,
    labSize = 6.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    colConnectors = 'black') +
  labs(subtitle = NULL)

Ptprc_volcano
```


# CD45- differential expression

```{r Structural differential expression}
Structural <- SetIdent(Structural, value = "treatment")

# Find the differential expressed genes in Structural
Structural_differential <- FindMarkers(Structural, ident.1 = "Alt.BPC", ident.2 = "Alt", min.cells.group = 10, test.use = "wilcox_limma", only.pos = FALSE, assay = "RNA", slot = "scale.data")

Structural_differential <- Structural_differential[order(abs(Structural_differential$avg_log2FC), decreasing = T), ]

head(Structural_differential)

# Find the top 5 (positive and negative) differential expression in ILC2
Structural_differential %>%
  filter(abs(avg_log2FC) > 0.5) %>%
  filter(avg_log2FC > 0) %>%
  top_n(-5, p_val) %>%
  arrange(p_val) -> Structural_differential_top5pos

head(Structural_differential_top5pos)

Structural_differential %>%
  filter(abs(avg_log2FC) > 0.5) %>%
  filter(avg_log2FC < 0) %>%
  top_n(-5, p_val) %>%
  arrange(p_val) -> Structural_differential_top5neg

head(Structural_differential_top5neg)
```

```{r Structural DE Volcano plot, fig.width = 6, fig.height = 6}
# Plot the top 10 differential genes with volcano plot
Structural_volcano <- EnhancedVolcano(Structural_differential,
    lab = rownames(Structural_differential),
    selectLab = c(rownames(Structural_differential_top5pos), rownames(Structural_differential_top5neg)),
    x = 'avg_log2FC',
    y = 'p_val',
    title = 'CD45- Alt.BPC vs. Alt',
    pCutoff = 10e-20,
    FCcutoff = 0.5,
    pointSize = 3.0,
    labSize = 6.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    colConnectors = 'black') +
  labs(subtitle = NULL)

Structural_volcano
```


# HSP gene expression

```{r HSP gene set}
# HSP genes
HSP <- c("Hsp1a1", "Hspa5", "Hspa8", "Hsp90aa1", "Hsp90ab1", "Hsp90b1")
```

```{r HSP gene expression, fig.width = 5, fig.height = 5}
# HSP expression in other leukocytes
HSP_Ptprc <- DotPlot(Ptprc, features = c(HSP), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("HSP genes in CD45+Thy1-")

HSP_Ptprc


# HSP expression in structural cells
HSP_Structural <- DotPlot(Structural, features = c(HSP), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("HSP genes in CD45-")

HSP_Structural
```


# Data input (ILC)

```{r Load SCT processed ILC data}
# Load ILC data

# ILC
ILC <- readRDS("CD45_Thy1_ILC_paper_ver.rds")


# Set RNA assay as default
DefaultAssay(ILC) <- "RNA"
head(x = ILC[[]])
```


# ILC cluster Visualization

```{r Set Reveal ILC clusters}
# Show the numbers for ILC
table(ILC@meta.data$cell_type, split.by = ILC@meta.data$treatment)

# Plot the ILC clusters
DimPlot(ILC, group.by = "cell_type", label = T, repel =  T)
DimPlot(ILC, group.by = "treatment", label = F) + ggtitle(NULL)
```


# HSP receptor gene expression

```{r HSP receptor gene set}
# HSP receptor genes
HSPR <- c("Olr1", "Scarf1", "Stab1", "Klrc1", "Klrc2", "Klrk1")
```

```{r HSP receptor gene expression, fig.width = 5, fig.height = 5}
# HSP expression in other leukocytes
HSPR_ILC <- DotPlot(ILC, features = c(HSPR), group.by = "cell_type", split.by = "treatment", assay = "RNA", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black"))) + ggtitle("HSP receptor genes in ILC")

HSPR_ILC
```


# Packages for transcription factor activity inference  analysis

```{r Load package for transcription factor activity analysis}
# Load the package (decoupleR)
library(decoupleR)

# Only needed for data handling and plotting
library(pheatmap)
```


# TF activity inference with Weighted Mean

```{r Load CollecTRI network}
# Load the database of TFs and their transcriptional targets
net <- get_collectri(organism='mouse', split_complexes=FALSE)

head(net)
```

```{r Calculate the pseudotime on averall ILC}
# Run the Weighted Mean method (wmean)
# Extract the normalized log-transformed counts
ILC.mat <- as.matrix(ILC@assays$RNA@data)

# Run wmean
ILC.acts <- run_wmean(mat=ILC.mat, net=net, .source='source', .target='target',
                  .mor='mor', times = 100, minsize = 5)
head(ILC.acts)

# Extract norm_wmean and store it in tfswmean
ILC[['tfswmean']] <- ILC.acts %>%
  filter(statistic == 'norm_wmean') %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)

# Change assay
DefaultAssay(object = ILC) <- "tfswmean"

# Scale the data (ILC)
ILC <- ScaleData(ILC)
ILC@assays$tfswmean@data <- ILC@assays$tfswmean@scale.data
```


# Exploration of TF activities

```{r ILC Ahr activity and epxression , fig.width = 20, fig.height = 4}
ILC_Ahr_p1 <- DimPlot(ILC, reduction = "umap", group.by = "cell_type", label = TRUE, repel = TRUE, label.size = 3.5, pt.size = 0.5) + 
  NoLegend() + ggtitle('NK/ILC')

ILC_Ahr_p2 <- (FeaturePlot(ILC, features = c("Ahr"), split.by = "treatment") & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  theme(legend.position = c(1, 0.5))

DefaultAssay(object = ILC) <- "RNA"

ILC_Ahr_p3 <- FeaturePlot(ILC, features = c("Ahr"), split.by = "treatment") +
  theme(legend.position = c(0.99, 0.5))

DefaultAssay(object = ILC) <- "tfswmean"

plot_grid(NULL, NULL, NULL, ILC_Ahr_p1, ILC_Ahr_p2, ILC_Ahr_p3, nrow = 2, ncol = 3, rel_heights = c(0.1, 1), rel_widths = c(1, 2, 2), labels = c('Cluster layout', 'AhR activity', 'Ahr expression'), label_size =16)

ILC_Ahr <- plot_grid(NULL, NULL, ILC_Ahr_p1, ILC_Ahr_p2, nrow = 2, ncol = 2, rel_heights = c(0.1, 1), rel_widths = c(1, 2))

ILC_Ahr
```

```{r ILC Ahr response gene expression , fig.width = 4.5, fig.height = 5}
# AhR genes in ILCs
ILC_Ahr_gene <- DotPlot(ILC, features = c("Cyp1a1", "Cyp1b1", "Nfe2l2", "Klf4"), assay ="RNA", group.by = "cell_type", split.by = "treatment", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black")))

ILC_Ahr_gene + ggtitle("AhR response genes")
```

```{r ILC Ahr TF activity, fig.width = 4, fig.height = 5}
# AhR downstream TFs
# https://link.springer.com/article/10.1007/s00204-019-02620-5

# AhR activity in ILCs
ILC_Ahr_TF <- DotPlot(ILC, features = c("Ahr", "Nfe2l2", "Klf4"), group.by = "cell_type", split.by = "treatment", cols = "RdBu") + theme(axis.title = element_blank(), legend.title = element_text(size = 8, face ="bold"), legend.text = element_text(size = 8), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), axis.text.y = element_text(color = c("dimgray", "black")))

ILC_Ahr_TF + ggtitle("Inferred AhR pathway activity")
```


# Data oupput

```{r HSP genes for paper, fig.width = 12, fig.height = 16}
# Generate the plot in paper figure form
figure7_1 <- plot_grid(Ptprc_volcano, NULL, HSP_Ptprc, labels = c('(A)', '(C)'), label_size = 18, nrow = 1, ncol = 3, rel_widths = c(1.25, 0.05, 1))
figure7_2 <- plot_grid(Structural_volcano, HSP_Structural, labels = c('(B)', '(D)'), label_size = 18, nrow = 1, ncol = 2, rel_widths = c(1.2, 1))
figure7_3 <- plot_grid(HSPR_ILC, labels = c('(E)'), label_size = 18, nrow = 1, ncol = 2, rel_widths = c(0.8, 1))


figure7 <- plot_grid(figure7_1, figure7_2, figure7_3,
                     nrow = 3,
                     ncol = 1,
                     rel_heights = c(1.2, 1.2, 1.1)
  )

figure7
```

```{r AhR activity for paper, fig.width = 10, fig.height = 9}
# Generate the plot in paper figure form
figure8_1 <- plot_grid(NULL, ILC_Ahr_gene, labels = c('(A)            AhR response genes'), label_x = -0.3, label_size = 18, nrow = 2, ncol = 1, rel_heights = c(0.1, 1))
figure8_2 <- plot_grid(NULL, ILC_Ahr_TF, labels = c('(B)       Inferred AhR pathway activity'), label_x = -0.40, label_size = 18, nrow = 2, ncol = 1, rel_heights = c(0.1, 1))
figure8_3 <- plot_grid(figure8_1, figure8_2, nrow = 1)
figure8_4 <- plot_grid(NULL, ILC_Ahr, labels = c('(C)   Inferred AhR pathway activity'), label_x = -0.2, label_size = 18, nrow = 2, ncol = 1, rel_heights = c(0.1, 1))

figure8 <- plot_grid(figure8_3, figure8_4,
                     nrow = 2,
                     ncol = 1,
                     rel_heights = c(1.4, 1)
  )

figure8
```


# Data save

```{r Output the figures}
# Create a folder for the figures
dir.create("./Figures", showWarnings = FALSE)

# Save the figure
ggsave2(plot = figure7, path = "./Figures", "Figure7.tiff", device = "tiff", dpi = 300, bg = "white", width = 12, height = 16, units = "in")
ggsave2(plot = figure8, path = "./Figures", "Figure8.tiff", device = "tiff", dpi = 300, bg = "white", width = 10, height = 9, units = "in")
```


<details>
  <summary>**Session Info**</summary>
```{r sessionInfo}
sessionInfo()
```
</details>

